library/common/http/dispatcher.cc

```diff
@@ -425,63 +412,31 @@ envoy_status_t Dispatcher::sendTrailers(envoy_stream_t stream, envoy_headers tra
-          // This interaction is important. The runResetCallbacks call synchronously causes Envoy to
-          // defer delete the HCM's ActiveStream. That means that the lifetime of the DirectStream
-          // only needs to be as long as that deferred delete. Therefore, we synchronously call
-          // cleanup here which will defer delete the DirectStream, which by definition will be
-          // scheduled **after** the HCM's defer delete as they are scheduled on the same dispatcher
-          // context.
-          //
-          // StreamResetReason::RemoteReset is used as the platform code that issues the
-          // cancellation is considered the remote.
-          //
-          // This call is guarded by hcm_stream_pending_destroy_ to protect against the
-          // following race condition:
-          //   1. resetStream executes first on a platform thread, getting through the dispatch
-          //   guard and posting this lambda.
-          //   2. The event dispatcher's thread executes a terminal encoding or a reset in the
-          //   Http::ConnectionManager, thus calling deferredDelete on the ActiveStream.
-          //   3. The event dispatcher's thread executes this post body, thus calling
-          //   runResetCallbacks, which ends up calling deferredDelete (for a second time!) on the
-          //   ActiveStream.
-          // This protection makes sure that Envoy Mobile's Http::Dispatcher::DirectStream knows
-          // synchronously when the ActiveStream is deferredDelete'd for the first time.
-          if (!direct_stream->hcm_stream_pending_destroy_) {
-            direct_stream->hcm_stream_pending_destroy_ = true;
-            direct_stream->runResetCallbacks(StreamResetReason::RemoteReset);
-          }
-          cleanup(direct_stream->stream_handle_);
-        }
-      });
+envoy_status_t Dispatcher::cancelStream(envoy_stream_t stream) {
+  post([this, stream]() -> void {
+    Dispatcher::DirectStreamSharedPtr direct_stream = getStream(stream);
+    if (direct_stream) {
+      Dispatcher::checkGarbage(direct_stream.get());
+      if (direct_stream->dispatchable(true)) {
+        direct_stream->callbacks_->onCancel();
+      }
+      // This interaction is important. The runResetCallbacks call synchronously causes Envoy to
+      // defer delete the HCM's ActiveStream. That means that the lifetime of the DirectStream
+      // only needs to be as long as that deferred delete. Therefore, we synchronously call
+      // cleanup here which will defer delete the DirectStream, which by definition will be
+      // scheduled **after** the HCM's defer delete as they are scheduled on the same dispatcher
+      // context.
+      //
+      // StreamResetReason::RemoteReset is used as the platform code that issues the
+      // cancellation is considered the remote.
+      if (!direct_stream->hcm_stream_pending_destroy_) {
+        direct_stream->hcm_stream_pending_destroy_ = true;
```

Comment:
I still don't understand why this boolean is necessary. Once a stream is pending destroy, it should be gone from the map and the map lookup on stream ID should fail. Can you try to remove this boolean and simplify the logic further?

Reply:
I actually was thinking through whether it was stlil necessary last night. I wasn't 100% convinced it was removable and I erred on the side of changing less, but I'll walk through the paths that use it.